[fastlane] Fix S3ClientHelper side effects

atreat opened the pull request:
<!-- Thanks for contributing to _fastlane_! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->

### Checklist
- [x] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
- [x] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
- [x] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary.

### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->

When testing the changes in https://github.com/fastlane/fastlane/pull/16682 I found that my authentication was not working because the S3ClientHelper was overwriting the Aws config I expected. This change will remove that side-effect and allow the caller to dictate what AWS credentials should be used.

### Description
<!-- Describe your changes in detail. -->
<!-- Please describe in detail how you tested your changes. -->

Updates the S3ClientHelper to remove the access_key/secret initializer parameters. Removes the call to `Aws.config.update`. Instead, passes the remaining `region` parameter to the S3 client when initializing that.

Also cleans up some confusing code in the file. The `client` method was defined as private outside of the S3ClientHelper class. The S3ClientHelper then had a public `attr_reader :client` that allowed the underlying client to leak out. This has been updated such that there is a private `client` method that is not accessible outside the helper class as I believe the intention is to always use the interface defined by the helper.

### Testing Steps
<!-- Optional: steps, commands, or code used to test your changes. -->
<!-- Providing these will reduce the time needed for testing and review by the fastlane team. -->

Updated the necessary call sites in match's S3Storage and specs. Specs for the S3ClientHelper continue to pass.

Commit:
Updates S3ClientHelper to prevent side effects

The initializer was calling Aws.config.update which was overwriting
existing configuration. This was preventing access when using
STS/Aws::AssumeRoleCredentials to assume a temporary IAM role.

Removing the need to initialize with access_key/secret allows the caller
to update Aws.config before calling fastlane and having their expected
credentials used. Credentials will now be pulled from Amazon's
documented sources.

Maintains `region` option by initializing the S3 client with this
option.

Commit:
Updates match's S3Storage object to initialize S3ClientHelper with new
interface.

Commit:
Updates S3ClientHelper with option s3_client initializer option

This is mainly to be used to inject this dependency in for specs, but if
callers feel they would rather configure an s3 client this allows them
to do so.

janpio commented:
Could this be a breaking change to anyone or is this deep enough in the code that we do not think anyone uses this directly?

atreat commented:
I couldn't find any other callers of the S3ClientHelper. It is possible that people call this directly though. I can add back in the original initializer and print out a deprecation notice if that is best.

Commit:
Updates S3ClientHelper with much more versatile initializer

Reverts changes to keep initializer interface non-breaking.
Updates creation of underlying S3 client to support passed in
credentials (access_key/secret). Will use AWS config otherwise

atreat commented:
Realized I could pass the access_key/secret to the underlying S3 client if they exist. This allows for the existing interface to be used and removes the need to call `Aws.config`

atreat commented:
@janpio let me know if you have any feedback on the updates

joshdholtz commented on fastlane/lib/fastlane/helper/s3_client_helper.rb:
```diff
@@ -50,12 +51,26 @@ def find_bucket!(bucket_name)
+      end
+
+      def create_credentials
+        return nil if access_key.nil? || @secret_access_key.nil?
```
Is there reason why `secret_access_key` has the `@` in front of it and `access_key` doesn't? 🤔 Just asking so we can have consistency or a comment mentioning why 😇

atreat replied:
This is a small attempt at keeping the secret_access_key from being accessed using an attr_reader. I can make that private or if its not a concern I can just open it up alongside the access_key.

joshdholtz replied:
I don't think its too much of a concern! I would go for consistency here if possible 😊

joshdholtz commented on fastlane/lib/fastlane/helper/s3_client_helper.rb:
```diff
@@ -50,12 +51,26 @@ def find_bucket!(bucket_name)
+
+        Aws::Credentials.new(
+          access_key,
+          @secret_access_key
```
Same here ☝️

joshdholtz commented on match/lib/match/storage/s3_storage.rb:
```diff
@@ -45,16 +41,14 @@ def self.configure(params)
                      team_name: nil)
         @s3_bucket = s3_bucket
         @s3_region = s3_region
-        @s3_client = Fastlane::Helper::S3ClientHelper.new(access_key: s3_access_key, secret_access_key: s3_secret_access_key, region: s3_region)
```
What is the reason for removing the access_key and secret_key from here? Isn't that needed for authenticating? I _think_ that access key and secret is needed for match to authenticate the user Into S3?

atreat replied:
I think I just missed revisiting this after adding back in the access_key/secret. Will update.

joshdholtz reviewed (changes requested):
I like this change! Just a few small changes requested / questions 😊

Commit:
Reverts change in S3Storage that prevented access_key/secret from being
used if passed in

joshdholtz reviewed (approved):
Thanks for making those changes! This looks 🔥 Will get this released in tonight's release 💪
